src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java

```diff
@@ -541,6 +698,31 @@ private boolean isTaskRequired(List<Map<String, CamundaVariable>> evaluateDmnRes
 
     }
 
+    /**
+     * Helper method to evaluate whether a user should have access to a task.
+     * If the user does not have access it will throw a {@link RoleAssignmentVerificationException}
+     *
+     * @param taskId                the task id obtained from camunda.
+     * @param accessControlResponse the access control response containing user's role assignment.
+     * @param permissionsRequired   the permissions that are required by the endpoint.
+     */
+    private TaskResource roleAssignmentVerification(String taskId,
+                                            AccessControlResponse accessControlResponse,
+                                            List<PermissionTypes> permissionsRequired) {
+        Optional<TaskResource> optionalTaskResource = cftQueryService
+            .getTask(taskId, accessControlResponse, permissionsRequired);
+
+        if (optionalTaskResource.isEmpty()) {
+            Optional<TaskResource> optionalTask = cftTaskDatabaseService.findByIdOnly(taskId);
+            if (optionalTask.isEmpty()) {
+                throw new TaskNotFoundException(TASK_NOT_FOUND_ERROR);
+            } else {
+                throw new RoleAssignmentVerificationException(ROLE_ASSIGNMENT_VERIFICATIONS_FAILED);
+            }
+        }
+        return optionalTaskResource.get();
```

Comment:
I checked this logic with Ravi because most of the time will exist so we don't check the existence everytime.
only when the permission checks fails, we check if the task id actually exists